From dbfe424164c884ff57ab0528bb1aa6bb3d7a5b8b Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Mon, 19 Oct 2026 10:28:36 -0700
Subject: [PATCH] test: add mine_mature_coin() for the tests' mine-and-mature
 blocks

18 Chronik tests began with the same lines: mine a block to the
anyone-can-spend P2SH address, look up its coinbase, mine
COINBASE_MATURITY blocks to an unspendable address, and hard-code the
coinbase's value. They now call self.mine_mature_coin(node), which returns
the coinbase's txid and its value as reported by the node.

Tests that need the tip after maturing call node.getbestblockhash().
Nine tests whose blocks differ (they use all the maturing blocks' hashes,
query Chronik between the two steps, or mature to another address) are
left as they are.
---
 doc/functional-tests.md                       |  4 ++++
 test/functional/chronik_block_info.py         | 13 +++--------
 test/functional/chronik_block_txs.py          | 13 ++---------
 .../functional/chronik_disable_token_index.py | 10 +--------
 test/functional/chronik_lokad_id_group.py     | 10 ++-------
 test/functional/chronik_pause.py              | 12 ++--------
 test/functional/chronik_raw_tx.py             | 10 ++-------
 .../chronik_script_batch_summary.py           |  8 ++-----
 test/functional/chronik_spent_by.py           | 13 +++--------
 .../functional/chronik_token_broadcast_txs.py | 11 +---------
 test/functional/chronik_token_burn.py         | 10 +--------
 test/functional/chronik_token_id_group.py     | 10 +--------
 .../functional/chronik_token_parse_failure.py | 10 +--------
 test/functional/chronik_token_script_group.py | 10 +--------
 test/functional/chronik_token_slp_fungible.py | 10 +--------
 test/functional/chronik_token_slp_nft1.py     | 10 +--------
 test/functional/chronik_tx.py                 |  9 +-------
 test/functional/chronik_ws_avalanche.py       | 11 ++--------
 test/functional/chronik_ws_ordering.py        |  9 +-------
 .../test_framework/test_framework.py          | 22 ++++++++++++++++++-
 20 files changed, 53 insertions(+), 162 deletions(-)

diff --git a/doc/functional-tests.md b/doc/functional-tests.md
index ed662e9110fd933377078b0d476475e96920a4b4..4fab9363f032ef03808ea702c1e91e5d7fca2bee 100644
--- a/doc/functional-tests.md
+++ b/doc/functional-tests.md
@@ -401,6 +401,10 @@ class WalletExampleTest(BitcoinTestFramework):
   to the test log, e.g. `self.log.info('Create enough transactions to fill a block')`.
   Logs make the test code easier to read and the test logic easier
   [to debug](#test-logging).
+- For one spendable coin, use `cointx, coinvalue = self.mine_mature_coin(node)`:
+  a coinbase paying to `ADDRESS_ECREG_P2SH_OP_TRUE` (spend it with
+  `SCRIPTSIG_OP_TRUE`), matured by `COINBASE_MATURITY` more blocks. For many
+  coins, use `MiniWallet`.
 - Set `self.num_nodes` to the minimum number of nodes necessary for the test.
   Having additional unrequired nodes adds to the execution time of the test as
   well as memory/CPU/disk requirements (which is important when running tests in
diff --git a/test/functional/chronik_block_info.py b/test/functional/chronik_block_info.py
index e31f0fcf0b310a5e268ae46494beb3ae98dc00d2..65c83c78ce4b8b32685f19166a703351d05af695 100644
--- a/test/functional/chronik_block_info.py
+++ b/test/functional/chronik_block_info.py
@@ -6,12 +6,11 @@ Test if the `BlockInfo` fields are set correctly in Chronik.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY, create_block, create_coinbase
+from test_framework.blocktools import create_block, create_coinbase
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
 from test_framework.script import OP_RETURN, CScript
@@ -34,15 +33,9 @@ class ChronikBlockInfoTest(BitcoinTestFramework):
 
         peer = node.add_p2p_connection(P2PDataStore())
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
+        cointx, coinvalue = self.mine_mature_coin(node)
+        prev_hash = node.getbestblockhash()
 
-        prev_hash = self.generatetoaddress(
-            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE
-        )[-1]
-
-        coinvalue = 5000000000
         tx = CTransaction()
         tx.vin = [
             CTxIn(outpoint=COutPoint(int(cointx, 16), 0), scriptSig=SCRIPTSIG_OP_TRUE)
diff --git a/test/functional/chronik_block_txs.py b/test/functional/chronik_block_txs.py
index 908bf9876e50d5c4088e6bac6bc6931d5093f248..d97f268fea552057ee2abc5a3fca8c89f4e1f14f 100644
--- a/test/functional/chronik_block_txs.py
+++ b/test/functional/chronik_block_txs.py
@@ -6,13 +6,10 @@ Test Chronik's /block-txs/:hash_or_height endpoint.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
-    ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.blocktools import (
-    COINBASE_MATURITY,
     GENESIS_BLOCK_HASH,
     create_block,
     create_coinbase,
@@ -92,15 +89,9 @@ class ChronikBlockTxsTest(BitcoinTestFramework):
             ),
         )
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
+        cointx, coinvalue = self.mine_mature_coin(node)
+        tip = node.getbestblockhash()
 
-        tip = self.generatetoaddress(
-            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE
-        )[-1]
-
-        coinvalue = 5000000000
         tx1 = CTransaction()
         tx1.vin = [
             CTxIn(outpoint=COutPoint(int(cointx, 16), 0), scriptSig=SCRIPTSIG_OP_TRUE)
diff --git a/test/functional/chronik_disable_token_index.py b/test/functional/chronik_disable_token_index.py
index ce0b695f1fd2729a1b55d7e4493bf62413d58f01..949a96c742da0c083ad74a45dfa0cc01d99003d5 100644
--- a/test/functional/chronik_disable_token_index.py
+++ b/test/functional/chronik_disable_token_index.py
@@ -6,12 +6,10 @@ Test Chronik's -chroniktokenindex=0 works properly.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.alp import alp_genesis, alp_opreturn, alp_send
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
@@ -40,13 +38,7 @@ class ChronikDisableTokenIndex(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         # ALP GENESIS tx
         tx = CTransaction()
diff --git a/test/functional/chronik_lokad_id_group.py b/test/functional/chronik_lokad_id_group.py
index 7bdd8bd5c5f7f1bb5033e4c3a065f0f6874ce41f..f75ddbe5161c4cf401f0d879395e989003db2d8c 100644
--- a/test/functional/chronik_lokad_id_group.py
+++ b/test/functional/chronik_lokad_id_group.py
@@ -7,11 +7,10 @@ Test Chronik indexes tx by LOKAD ID correctly.
 
 from test_framework.address import (
     ADDRESS_ECREG_P2SH_OP_TRUE,
-    ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY, create_block, create_coinbase
+from test_framework.blocktools import create_block, create_coinbase
 from test_framework.hash import hash160
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -39,12 +38,7 @@ class ChronikLokadIdGroup(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-        coinvalue = 5000000000
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         def p2lokad(lokad_id: bytes):
             return CScript([lokad_id, OP_EQUAL])
diff --git a/test/functional/chronik_pause.py b/test/functional/chronik_pause.py
index dc245f03bc8a84430a45f4c4d488c6b0d7c54f5f..86bf853d045100b76ec73c30ede3371ceafca5f4 100644
--- a/test/functional/chronik_pause.py
+++ b/test/functional/chronik_pause.py
@@ -5,13 +5,10 @@
 import time
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
-    ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.blocktools import (
-    COINBASE_MATURITY,
     create_block,
     create_coinbase,
     make_conform_to_ctor,
@@ -45,13 +42,8 @@ class ChronikPauseTest(BitcoinTestFramework):
         peer = node.add_p2p_connection(P2PDataStore())
 
         self.log.info("Generate coin + mature for testing")
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-        coinvalue = 5000000000
-        tip = self.generatetoaddress(
-            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE
-        )[-1]
+        cointx, coinvalue = self.mine_mature_coin(node)
+        tip = node.getbestblockhash()
 
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointx, 16), 0), SCRIPTSIG_OP_TRUE)]
diff --git a/test/functional/chronik_raw_tx.py b/test/functional/chronik_raw_tx.py
index d5bdd6061db2ac65506246c7215477a5aa9739d8..2cf22d4e018ef472b4d433b7767f4b9e60f53988 100644
--- a/test/functional/chronik_raw_tx.py
+++ b/test/functional/chronik_raw_tx.py
@@ -6,11 +6,10 @@ Test Chronik's /raw-tx/:txid endpoint.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY, GENESIS_CB_TXID
+from test_framework.blocktools import GENESIS_CB_TXID
 from test_framework.hash import hash160
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.script import OP_EQUAL, OP_HASH160, CScript
@@ -64,13 +63,8 @@ class ChronikRawTxTest(BitcoinTestFramework):
             ),
         )
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
+        cointx, coinvalue = self.mine_mature_coin(node)
 
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
         send_values = [coinvalue - 10000, 1000, 2000, 3000]
         send_redeem_scripts = [bytes([i + 0x52]) for i in range(len(send_values))]
         send_scripts = [
diff --git a/test/functional/chronik_script_batch_summary.py b/test/functional/chronik_script_batch_summary.py
index 7cdf82eafdb32b8db5b6ab6e42c614c5a1f22d18..bf031caa285a3a7bd8ccf8f2c3a7d47ee35a59d6 100644
--- a/test/functional/chronik_script_batch_summary.py
+++ b/test/functional/chronik_script_batch_summary.py
@@ -6,12 +6,11 @@ Test Chronik POST /script/batch/summary (ScriptBatchSummary protos).
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY, GENESIS_CB_PK
+from test_framework.blocktools import GENESIS_CB_PK
 from test_framework.chronik.alp import alp_genesis, alp_opreturn
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.hash import hash160
@@ -83,11 +82,8 @@ class ChronikScriptBatchSummaryTest(BitcoinTestFramework):
 
         # P2SH (OP_TRUE redeem) plus a P2PKH output that holds an ALP token UTXO
         node.setmocktime(1300000000)
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        cointx = node.getblock(coinblockhash)["tx"][0]
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
+        cointx, coinvalue = self.mine_mature_coin(node)
 
-        coinvalue = 5000000000
         pubkey, token_spk, _token_addr = getnewdestination()
         token_keyhash_hex = hash160(pubkey).hex()
 
diff --git a/test/functional/chronik_spent_by.py b/test/functional/chronik_spent_by.py
index 443cd888fbecdef07b3469ebf62d198f1ddcf180..b40dbd7f819b295328a58d3e3bfc3901b1a18e39 100644
--- a/test/functional/chronik_spent_by.py
+++ b/test/functional/chronik_spent_by.py
@@ -6,11 +6,10 @@ Test if the `Tx.spent_by` field is set correctly in Chronik.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY, create_block, create_coinbase
+from test_framework.blocktools import create_block, create_coinbase
 from test_framework.hash import hash160
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -35,15 +34,9 @@ class ChronikSpentByTest(BitcoinTestFramework):
 
         peer = node.add_p2p_connection(P2PDataStore())
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
+        cointx, coinvalue = self.mine_mature_coin(node)
+        tip = node.getbestblockhash()
 
-        tip = self.generatetoaddress(
-            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE
-        )[-1]
-
-        coinvalue = 5000000000
         send_values = [coinvalue - 10000, 1000, 1000, 1000]
         send_redeem_scripts = [bytes([i + 0x52]) for i in range(len(send_values))]
         send_script_hashes = [hash160(script) for script in send_redeem_scripts]
diff --git a/test/functional/chronik_token_broadcast_txs.py b/test/functional/chronik_token_broadcast_txs.py
index 7d5fde9ba0b3d0935e58eef3bc7c2aec708b21ad..03e575cf1062e47aa2e3480b88b511610066797c 100644
--- a/test/functional/chronik_token_broadcast_txs.py
+++ b/test/functional/chronik_token_broadcast_txs.py
@@ -6,12 +6,9 @@ Test Chronik broadcasts a batch of txs correctly.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
-    ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.alp import alp_genesis, alp_opreturn, alp_send
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
@@ -39,13 +36,7 @@ class ChronikTokenBroadcastTxs(BitcoinTestFramework):
         node = self.nodes[0]
         chronik = node.get_chronik_client()
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_token_burn.py b/test/functional/chronik_token_burn.py
index 1e81b57fcb82bddba66bb0a3d174409752e96872..c448d6906ef6c0153dbd483c646385a6f2516575 100644
--- a/test/functional/chronik_token_burn.py
+++ b/test/functional/chronik_token_burn.py
@@ -6,12 +6,10 @@ Test Chronik indexes token burns correctly.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.alp import alp_burn, alp_genesis, alp_opreturn, alp_send
 from test_framework.chronik.slp import slp_burn, slp_genesis
 from test_framework.chronik.token_tx import TokenTx
@@ -47,13 +45,7 @@ class ChronikTokenBurn(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         # Make a normal non-token tx
         non_token_tx = CTransaction()
diff --git a/test/functional/chronik_token_id_group.py b/test/functional/chronik_token_id_group.py
index 243b2a6eec08999c4e3dd4bb71e8e620094f1e96..c36197ab297b9b0cfe7775949d4f653506e16b8b 100644
--- a/test/functional/chronik_token_id_group.py
+++ b/test/functional/chronik_token_id_group.py
@@ -8,12 +8,10 @@ Test Chronik indexes tx by token ID correctly.
 from itertools import zip_longest
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.alp import alp_genesis, alp_opreturn, alp_send
 from test_framework.chronik.slp import slp_genesis
 from test_framework.chronik.token_tx import TokenTx
@@ -60,13 +58,7 @@ class ChronikTokenBurn(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_token_parse_failure.py b/test/functional/chronik_token_parse_failure.py
index 09f898c6cd274b5ec81820dcf54e2f4631a1874b..5efa5c72cc3c34cc25474802f898d99b0df8ce68 100644
--- a/test/functional/chronik_token_parse_failure.py
+++ b/test/functional/chronik_token_parse_failure.py
@@ -6,12 +6,10 @@ Test Chronik indexes parse failures correctly.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.script import OP_RESERVED, OP_RETURN, CScript
@@ -34,13 +32,7 @@ class ChronikTokenParseFailure(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_token_script_group.py b/test/functional/chronik_token_script_group.py
index 33909431eba059690a8ea22186297cc9e7116b92..66e051782538c16165ab2ced0c37ef6fec0762c7 100644
--- a/test/functional/chronik_token_script_group.py
+++ b/test/functional/chronik_token_script_group.py
@@ -8,12 +8,10 @@ Test Chronik's /script endpoint works well with tokens.
 from itertools import zip_longest
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.alp import alp_genesis, alp_opreturn
 from test_framework.chronik.slp import slp_genesis, slp_send
 from test_framework.chronik.token_tx import TokenTx
@@ -51,13 +49,7 @@ class ChronikTokenScriptGroup(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_token_slp_fungible.py b/test/functional/chronik_token_slp_fungible.py
index f2effa1ef6e9aad9d7508a3e03e5a8bdd122dcd0..7ca9740aed33ae44f5e7e91a16103d06c5bd165b 100644
--- a/test/functional/chronik_token_slp_fungible.py
+++ b/test/functional/chronik_token_slp_fungible.py
@@ -6,12 +6,10 @@ Test Chronik indexes fungible SLP tokens.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.slp import slp_genesis, slp_mint, slp_send
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
@@ -40,13 +38,7 @@ class ChronikTokenSlpFungible(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_token_slp_nft1.py b/test/functional/chronik_token_slp_nft1.py
index a01913ffd3bc69f8c3abaf9bc45f1af17fcaf65e..87da51e18b3ea75f1a5ed341d6d8102d2b707d40 100644
--- a/test/functional/chronik_token_slp_nft1.py
+++ b/test/functional/chronik_token_slp_nft1.py
@@ -6,12 +6,10 @@ Test Chronik indexes SLP NFT1 tokens correctly.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.slp import slp_genesis, slp_mint, slp_send
 from test_framework.chronik.token_tx import TokenTx
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
@@ -46,13 +44,7 @@ class ChronikTokenSlpNft1(BitcoinTestFramework):
         mocktime = 1300000000
         node.setmocktime(mocktime)
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
-
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
+        cointx, coinvalue = self.mine_mature_coin(node)
 
         txs = []
 
diff --git a/test/functional/chronik_tx.py b/test/functional/chronik_tx.py
index f46b41c0add25467d94fc9c67a8e438b65c33cea..c5232a55eb5edfdb9994af8c0470ce9541fbf4ff 100644
--- a/test/functional/chronik_tx.py
+++ b/test/functional/chronik_tx.py
@@ -6,13 +6,11 @@ Test Chronik's /tx endpoint.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.blocktools import (
-    COINBASE_MATURITY,
     GENESIS_CB_TXID,
     create_block,
     create_coinbase,
@@ -61,13 +59,8 @@ class ChronikTxTest(BitcoinTestFramework):
         # Verify queried genesis tx matches
         assert_equal(chronik.tx(GENESIS_CB_TXID).ok(), genesis_cb_tx())
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointx = coinblock["tx"][0]
+        cointx, coinvalue = self.mine_mature_coin(node)
 
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
         send_values = [coinvalue - 10000, 1000, 2000, 3000]
         send_redeem_scripts = [bytes([i + 0x52]) for i in range(len(send_values))]
         send_scripts = [
diff --git a/test/functional/chronik_ws_avalanche.py b/test/functional/chronik_ws_avalanche.py
index 2115d01bf7db00c1c416f2c52d90918283e23fc7..e7f5db8f595590a0ca93807a3e4a2771fc36e7de 100644
--- a/test/functional/chronik_ws_avalanche.py
+++ b/test/functional/chronik_ws_avalanche.py
@@ -7,13 +7,10 @@ import os
 import time
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
-    ADDRESS_ECREG_UNSPENDABLE,
     SCRIPT_UNSPENDABLE,
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.avatools import can_find_inv_in_poll, get_ava_p2p_interface
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.chronik.slp import slp_genesis
 from test_framework.messages import (
     AvalancheTxVoteError,
@@ -136,12 +133,8 @@ class AvatestPlugin(Plugin):
 
         from test_framework.chronik.client import pb
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        coinbase_txid = coinblock["tx"][0]
-        tip = self.generatetoaddress(
-            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE
-        )[-1]
+        coinbase_txid, _ = self.mine_mature_coin(node)
+        tip = node.getbestblockhash()
 
         def has_finalized_tip(tip_expected):
             hash_tip_final = int(tip_expected, 16)
diff --git a/test/functional/chronik_ws_ordering.py b/test/functional/chronik_ws_ordering.py
index 05b821583b59f95c95754ee33a84e493f84a0f32..5ccc8ea6005986e8d36166ebeb5c85fe365e947c 100644
--- a/test/functional/chronik_ws_ordering.py
+++ b/test/functional/chronik_ws_ordering.py
@@ -6,13 +6,11 @@ Test Chronik whether WS are emitted in deterministic order.
 """
 
 from test_framework.address import (
-    ADDRESS_ECREG_P2SH_OP_TRUE,
     ADDRESS_ECREG_UNSPENDABLE,
     P2SH_OP_TRUE,
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.avatools import AvaP2PInterface, can_find_inv_in_poll
-from test_framework.blocktools import COINBASE_MATURITY
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
 from test_framework.script import OP_CHECKSIG, CScript
@@ -49,13 +47,8 @@ class ChronikWsOrdering(BitcoinTestFramework):
         node.add_p2p_connection(P2PDataStore())
         chronik = node.get_chronik_client()
 
-        coinblockhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        coinblock = node.getblock(coinblockhash)
-        cointxid = coinblock["tx"][0]
+        cointxid, coinvalue = self.mine_mature_coin(node)
 
-        self.generatetoaddress(node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE)
-
-        coinvalue = 5000000000
         balance = coinvalue
         current_txid = cointxid
 
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 623aee1485d77c8b1acdad262e232fec01b9915d..7d5081e57c0dc8b84fbbce9d00ff100c6e95156f 100644
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -24,9 +24,10 @@ from enum import Enum
 from typing import AbstractSet, Any, Dict, List, Optional, Sequence, Tuple, Union
 
 from . import coverage
-from .address import ADDRESS_ECREG_P2SH_OP_TRUE
+from .address import ADDRESS_ECREG_P2SH_OP_TRUE, ADDRESS_ECREG_UNSPENDABLE
 from .authproxy import JSONRPCException
 from .avatools import get_proof_ids
+from .messages import XEC
 from .p2p import NetworkThread
 from .test_node import TestNode
 from .util import (
@@ -1081,6 +1082,25 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         sync_fun() if sync_fun else self.sync_all()
         return blocks
 
+    def mine_mature_coin(self, node, **kwargs) -> Tuple[str, int]:
+        """Mine a block whose coinbase pays to ADDRESS_ECREG_P2SH_OP_TRUE, then
+        COINBASE_MATURITY blocks (to ADDRESS_ECREG_UNSPENDABLE) so that it can
+        be spent, with SCRIPTSIG_OP_TRUE. Returns the coinbase's txid and the
+        value of its output 0, in satoshis. kwargs go to generatetoaddress()
+        (e.g. sync_fun)."""
+        # Not imported at module level: blocktools reads the source tree on
+        # import, and the test runner imports this module before it is known
+        from .blocktools import COINBASE_MATURITY
+
+        blockhash = self.generatetoaddress(
+            node, 1, ADDRESS_ECREG_P2SH_OP_TRUE, **kwargs
+        )[0]
+        coinbase = node.getblock(blockhash, 2)["tx"][0]
+        self.generatetoaddress(
+            node, COINBASE_MATURITY, ADDRESS_ECREG_UNSPENDABLE, **kwargs
+        )
+        return coinbase["txid"], int(coinbase["vout"][0]["value"] * XEC)
+
     def sync_blocks(self, nodes=None, wait=1, timeout=60):
         """
         Wait until everybody has the same tip.
-- 
2.50.1 (Apple Git-155)

